src/main/java/net/neoforged/neoforge/transfer/ResourceHandlerUtil.java

```diff
@@ -0,0 +1,565 @@
+     * // Target
+     * IResourceHandler<FluidResource> target;
+     * Predicate<FluidResource> filter = resource -> resource.is(Fluids.WATER);
+     *
+     * // Move exactly one bucket in total, only of water:
+     * try (Transaction transaction = Transaction.open(TransactionContext.ROOT)) {
+     *     int waterMoved = ResourceHandlerUtil.move(source, target, filter, FluidType.BUCKET_VOLUME, transaction);
+     *     if (waterMoved == FluidType.BUCKET_VOLUME) {
+     *         // Only commit if exactly one bucket was moved.
+     *         transaction.commit();
+     *     }
+     *     //If committed, leaving this try-block will keep all changes.
+     * }
+     * }</pre>
+     *
+     * @param from        The source handler. May be null.
+     * @param to          The target handler. May be null.
+     * @param filter      The filter for transferred resources.
+     *                    Only resources for which this filter returns {@code true} will be transferred.
+     *                    This filter will never be tested with an empty resource, and filters are encouraged to throw an
+     *                    exception if this guarantee is violated.
+     * @param amount      The maximum amount that will be transferred.
+     * @param transaction The transaction this transfer is part of, or {@code null} if a transaction should be opened just for this transfer.
+     * @param <T>         The type of resources to move.
+     * @return The total amount of resources that was successfully transferred. This number is not necessarily for one resource, as we only pass in a filter. It is intended to be used to determine a raw number of resources moved.
+     * @throws IllegalStateException If no transaction is passed and a transaction is already active on the current thread.
+     */
+    public static <T extends IResource> int move(
+            @Nullable IResourceHandler<T> from,
+            @Nullable IResourceHandler<T> to,
+            Predicate<T> filter,
+            int amount,
+            @Nullable TransactionContext transaction) {
+        Objects.requireNonNull(filter, "Filter may not be null");
+        if (from == null || to == null) return 0;
+
+        try (Transaction subTransaction = TransactionManager.open(transaction)) {
+            int totalMoved = 0;
+            int size = from.size();
+
+            for (int index = 0; index < size; ++index) {
+                T fromResource = from.getResource(index);
+                if (fromResource.isEmpty() || !filter.test(fromResource)) continue;
+
+                // check how much can be extracted
+                int maxExtracted;
+                try (Transaction simulatedExtract = TransactionManager.open(subTransaction)) {
+                    maxExtracted = from.extract(index, fromResource, amount - totalMoved, simulatedExtract);
+                }
+
+                try (Transaction transferTransaction = TransactionManager.open(subTransaction)) {
```

Comment:
If the amount that can be extracted is zero, then we should just skip opening a transaction and attempting to insert/extract from it.